lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs

```diff
@@ -427,27 +391,252 @@ cliRetry tr msg processConfig = do
     isFail (st, _, _) = pure (st /= ExitSuccess)
     pol = limitRetriesByCumulativeDelay 30_000_000 $ constantDelay 1_000_000
 
-newtype PoolConfig = PoolConfig
-    { retirementEpoch :: Maybe EpochNo
+-- | The idea of what kind if pool we want to set up.
+data PoolRecipe = PoolRecipe
+    { pledgeAmt :: Integer
+    , index :: Int
+    , retirementEpoch :: Maybe EpochNo
       -- ^ An optional retirement epoch. If specified, then a pool retirement
       -- certificate will be published after the pool is initially registered.
+    , poolMetadata :: Aeson.Value
+    , delisted :: Bool
     }
     deriving (Eq, Show)
 
+-- | Represents the notion of a fully configured pool. All keys are known, but
+-- not necessarily exposed using this interface.
+data ConfiguredPool = ConfiguredPool
+    { operatePool
+        :: forall a. NodeParams -> (RunningNode -> IO a) -> IO a
+      -- ^ Precondition: the pool must first be registered.
+    , metadataUrl
+        :: Text
+    , recipe
+        :: PoolRecipe
+      -- ^ The 'PoolRecipe' used to create this 'ConfiguredPool'.
+    , registerViaTx :: RunningNode -> IO ()
+    }
+
+data PoolMetadataServer = PoolMetadataServer
+    { registerMetadataForPoolIndex :: Int -> Aeson.Value -> IO ()
+    , urlFromPoolIndex  :: Int -> String
+    }
+
+withPoolMetadataServer
+    :: Tracer IO ClusterLog
+    -> FilePath
+    -> (PoolMetadataServer -> IO a)
+    -> IO a
+withPoolMetadataServer tr dir action = do
+    let metadir = dir </> "pool-metadata"
+    createDirectoryIfMissing False metadir
+    withStaticServer metadir $ \baseURL -> do
+        let _urlFromPoolIndex i = baseURL </> metadataFileName i
+        action $ PoolMetadataServer
+            { registerMetadataForPoolIndex = \i metadata -> do
+                let metadataBytes = Aeson.encode metadata
+                BL8.writeFile (metadir </> (metadataFileName i)) metadataBytes
+                let hash = blake2b256 (BL.toStrict metadataBytes)
+                traceWith tr $
+                    MsgRegisteringPoolMetadata
+                        (_urlFromPoolIndex i)
+                        (B8.unpack $ hex hash)
+            , urlFromPoolIndex = _urlFromPoolIndex
+            }
+  where
+
+    metadataFileName :: Int -> FilePath
+    metadataFileName i = show i <> ".json"
+
+configurePools
+    :: Tracer IO ClusterLog
+    -> FilePath
+    -> PoolMetadataServer
+    -> [PoolRecipe]
+    -> IO [ConfiguredPool]
+configurePools tr dir metadataServer =
+    mapM (configurePool tr dir metadataServer)
+
+configurePool
```

Comment:
I like the work you've done here with the clearer types and easier-to-understand program flow :100: